keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,634 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a feature to protect Secrets/ConfigMaps while it is in use. Currently, users can delete a secret that is being used by other resources, like Pods, PVs, and VolumeSnapshots. This may have negative impact on the resouces using the secret and it may result in data loss.
+
+For example, one of the worst case scenarios is deletion of a Node Stage Secret for CSI volumes while it is still in-use.
+If it happens, a CSI driver for the volume can't unstage the volume due to the lack of the Secret,
+as a result, the volume can't be detached from the node and the volume can't even be deleted.
+This issue will easily happen if a PVC for the volume and a Secret for the volume exist in the same namespace, and a user requests to delete the namespace, which starts deletion of all the resources in the namespace.
+When the Secret is deleted before the PVC is deleted, this issue happens.
+
+Also, ConfigMaps can be deleted while it is in use, which may lead to an unexpected behavior in applications.
+
+Similar features for protecting PV and PVC already exist as [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md).
+
+
+## Motivation
+
+This feature aims to protect Secrets/ConfigMaps from being deleted while they are in-use.
+Secrets can be used by below ways:
+- From Pod:
+  - [Mounted as data volumes](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables)
+  - [Generic ephemeral volumes
+](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes) (can be handled as CSI PV below)
+- From Deployment:
```

Comment:
`PodSpec` is used in CRDs of 3d party Kubernetes extensions. IMO it's not possible to protect them from a referenced Secret to be deleted. Is it then worth adding the protection to in-tree objects like Deployments and ignoring others?

Reply:
Good question.

The reason that I added them is the discussion [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r698495299).
Therefore, we have below 4 choices for how we handle Pod:

(1) Protect only from Pod
(2) Protect from Pod and Pod-releated k8s official resources (Deployment, StatefulSet, DaemonSet, CronJob)
(3) Protect from Pod and all Pod-releated resources (May technically impossible)
(4) Don't protect from Pod at all

Actually, even without (2) and (3), (1) should save most of the cases (problem may happen only in the short period of time that Pod doesn't exist). However, we are still not sure if (1) is really expected by users in the first place, therefore we might be able to choose (4) for now, or dropping it from the scope of alpha and wait for feedback from users.

WDYT?